### PR TITLE
Tell cachet upon succesful cache

### PR DESCRIPTION
**Pull request by zipy124**
## What does this PR do?
Upon a succesfull cache we now tell cachet, this means our status script only has to update when its down, this makes a lot more sense going from the primary source instead of the secondary. 

## ✅ Pull Request checklist

- [ ] Is this code complete?
- [ ] Are tests done/modified?
- [ ] Are docs written for this PR? (if applicable)

## 🚨 Is this a breaking change for API clients?
Yes/No

## :squirrel: Deploy notes
Enviroment variable needs cachet token adding.

### CODE REVIEW

File: backend/uclapi/timetable/management/commands/update_gencache.py [hunk 1]
```diff
@@ -342,4 +343,13 @@ def handle(self, *args, **options):
         # to be run again in the future.
         self._redis.delete(cache_running_key)
 
+        url = "https://cachet.apps.uclapi.com/api/v1/components/2"
```
Comment (ChristopherHammond13) [inline backend/uclapi/timetable/management/commands/update_gencache.py]: Can we move this URL to settings too?

File: backend/uclapi/timetable/management/commands/update_gencache.py [hunk 2]
```diff
@@ -342,4 +343,13 @@ def handle(self, *args, **options):
...
+        cachet_token = settings.CACHET_KEY
+        cachet_format = "application/x-www-form-urlencoded"
+        payload_headers = {"X-Cachet-Token": cachet_token,
+                           "Content-Type": cachet_format}
+        payload = {"status": 1}
+        requests.request("PUT", url, data=payload, headers=payload_headers)
```
Comment (ChristopherHammond13) [inline backend/uclapi/timetable/management/commands/update_gencache.py]: It looks like we can just use `requests.put` which might be a bit neater?

File: backend/uclapi/uclapi/settings.py [hunk 3]
```diff
@@ -23,7 +23,7 @@
 
 # SECURITY WARNING: keep the secret key used in production secret!
 SECRET_KEY = os.environ.get("SECRET_KEY")
-
+CACHET_KEY = os.environ.get("CACHET_TOKEN")
```
Comment (ChristopherHammond13) [inline backend/uclapi/uclapi/settings.py]: Any reason we don't juse call this variable `CACHET_TOKEN`?

Comment (zipy124) [reply backend/uclapi/uclapi/settings.py]: Nope I'm just bad at naming things :man_shrugging:

File: backend/uclapi/timetable/management/commands/update_gencache.py (hunk 2 above)
Comment (zipy124) [reply backend/uclapi/timetable/management/commands/update_gencache.py]: Yep let me stick that in

File: backend/uclapi/timetable/management/commands/update_gencache.py (hunk 1 above)
Comment (zipy124) [reply backend/uclapi/timetable/management/commands/update_gencache.py]: yeh that probably makes a lot of sense

File: backend/uclapi/workspaces/occupeye/cache.py [hunk 4]
```diff
@@ -625,6 +625,17 @@ def feed_cache(self, full):
         )
         self._redis.set(last_modified_key, current_timestamp)
 
+        if not settings.DEBUG:
+            url = settings.CACHET_URL
```
Comment (ChristopherHammond13) [inline backend/uclapi/workspaces/occupeye/cache.py]:
```suggestion
```
Fix below

File: backend/uclapi/workspaces/occupeye/cache.py [hunk 5]
```diff
@@ -625,6 +625,17 @@ def feed_cache(self, full):
         )
         self._redis.set(last_modified_key, current_timestamp)
 
+        if not settings.DEBUG:
+            url = settings.CACHET_URL
+            url = url+"components/3"
```
Comment (ChristopherHammond13) [inline backend/uclapi/workspaces/occupeye/cache.py]:
```suggestion
            url = "{}/components/3".format(settings.CACHET_URL)
```
Might be prettier

Although again it might be good to extract out component 3 into a settings variable somewhere because that can change if we update Cachet.

File: backend/uclapi/timetable/management/commands/update_gencache.py [hunk 6]
```diff
@@ -341,5 +342,15 @@ def handle(self, *args, **options):
         # Cache has been run now, so we can delete the key to allow it
         # to be run again in the future.
         self._redis.delete(cache_running_key)
+        if not settings.DEBUG:
+            url = settings.CACHET_URL
+            url = url+"components/2"
```
Comment (ChristopherHammond13) [inline backend/uclapi/timetable/management/commands/update_gencache.py]: See below for opinions on this

File: backend/uclapi/common/cachet.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        if target_comp["status"] == 1:
+            incidents = cachet.Incidents(endpoint=ENDPOINT, api_token=API_TOKEN)
+            message = f"Gencache failed on "
+                      f"{settings.UCLAPI_DOMAIN_CURRENT}"
+                      f" with error: {repr(error_message)}"
+            new_incident = json.loads(incidents.post(name='Gencache failed',
+                                 message=message,
+                                 status=1,
+                                 component_id=i["id"],
+                                 component_status=4))
```
Comment (Feverfew) [inline backend/uclapi/common/cachet.py]:
```suggestion
            new_incident = json.loads(
                incidents.post(
                    name='Gencache failed',
                    message=message,
                    status=1,
                    component_id=i["id"],
                    component_status=4
                )
            )
```
Hound would help here I imagine. But in any case, my confusion here is the indentation of the inner-most arguments (after the first) seem off.

File: backend/uclapi/common/cachet.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import cachetclient.cachet as cachet
+import json
+from django.conf import settings
+
+ENDPOINT = settings.CACHET_URL
+API_TOKEN = settings.CACHET_TOKEN
```
[the remark this answers is not in the document]
Comment (zipy124) [reply backend/uclapi/common/cachet.py:6]: Lmao, i put them in because Chris requested they be globals smh

File: backend/uclapi/common/cachet.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+def create_incident(error_message, component_name):
+    try:
+        components = cachet.Components(endpoint=ENDPOINT, api_token=API_TOKEN)
+        comp_list = json.loads(components.get())
+        target_comp = None
+        for i in comp_list["data"]:
+            if i["name"] == component_name:
+                target_comp = i
+                break
+        if target_comp["status"] == 1:
```
[the remark this answers is not in the document]
Comment (zipy124) [reply backend/uclapi/common/cachet.py:17]: Ok, will add explicit calls to it!

File: backend/uclapi/workspaces/occupeye/cache.py [hunk 10]
```diff
@@ -641,6 +641,17 @@ def feed_cache(self, full):
...
+        if not settings.DEBUG:
+            url = settings.CACHET_URL
+            url = url+"components/3"
+            cachet_token = settings.CACHET_TOKEN
+            cachet_format = "application/x-www-form-urlencoded"
+            payload_headers = {"X-Cachet-Token": cachet_token,
+                               "Content-Type": cachet_format}
+            payload = {"status": 1}
+            requests.put(url, data=payload, headers=payload_headers)
+            print("[+] Cachet updated")
```
[the remark this answers is not in the document]
Comment (zipy124) [reply backend/uclapi/workspaces/occupeye/cache.py:653]: Printing is effectively logging, since it is redirected to a log file when run.

File: backend/uclapi/timetable/management/commands/update_gencache.py [hunk 11]
```diff
@@ -358,5 +363,9 @@ def handle(self, *args, **options):
             except Exception as teams_error:
                 print(f"Failed to send message to Microsoft Teams. "
                       f"Reason: {repr(teams_error)}")
+
+            if settings.UCLAPI_DOMAIN_CURRENT == "staging.ninja":
```
[the remark this answers is not in the document]
Comment (zipy124) [reply backend/uclapi/timetable/management/commands/update_gencache.py:367]: WIP, aparently I forgot to add the second else lmao.

File: backend/uclapi/common/cachet.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
+import cachetclient.cachet as cachet
+import json
+from django.conf import settings
+
+
+def create_incident(error_message, component_name):
+    try:
+        components = cachet.Components(
+            endpoint=settings.CACHET_URL, api_token=settings.CACHET_TOKEN)
```
[the remark this answers is not in the document]
Comment (zipy124) [reply backend/uclapi/common/cachet.py:9]: Figured out what option this is and disabled in hound, useless since we can't introspect more information out of flake8-black at this time.

File: backend/uclapi/workspaces/occupeye/cache.py [hunk 13]
```diff
@@ -641,6 +641,17 @@ def feed_cache(self, full):
         )
         self._redis.set(last_modified_key, current_timestamp)
 
+        if not settings.DEBUG:
+            url = settings.CACHET_URL
+            url = url+"components/3"
```
[the remark this answers is not in the document]
Comment (zipy124) [reply backend/uclapi/workspaces/occupeye/cache.py:646]: Lmao I forgot about this part ermmm, I'll change that to use the new code...

File: backend/uclapi/timetable/management/commands/update_gencache.py [hunk 14]
```diff
@@ -22,6 +22,10 @@
     Room, RoomA, RoomB, \
     Booking, BookingA, BookingB
 from common.helpers import LOCAL_TIMEZONE
+
+from common.cachet import create_incident, delete_incident, CachetException, \
+    get_incident_name
```
Comment (Feverfew) [inline backend/uclapi/timetable/management/commands/update_gencache.py:27]: Ughhh, it's annoying we hit the 80 line mark here. Looks a bit awkward, guess good opportunity for [PEP8 compliance](https://stackoverflow.com/questions/29534027/why-import-statements-with-parentheses) (can Hound do it's job please???)

File: backend/uclapi/common/cachet.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+import json
+from django.conf import settings
+
+
+class CachetException(Exception):
+    pass
+
+
+def get_incident_name(base):
+    incident_name = None
```
[the remark this answers is not in the document]
Comment (Feverfew) [reply backend/uclapi/common/cachet.py:11]: So, `None` may be more "correct" here, but empty string is "safer" IMO (as someone will be bound to assume it's always a string), and it's ok to use an empty-string in this case. However, feel free to ignore if you disagree, I don't have a strong opinion on this.

File: backend/uclapi/timetable/management/commands/update_gencache.py [hunk 16]
```diff
@@ -358,5 +377,17 @@ def handle(self, *args, **options):
             except Exception as teams_error:
                 print(f"Failed to send message to Microsoft Teams. "
                       f"Reason: {repr(teams_error)}")
+
+            incident_name = get_incident_name("Gencache")
+            if incident_name is not None:
```
Comment (Feverfew) [inline backend/uclapi/timetable/management/commands/update_gencache.py:382]:
```suggestion
            if incident_name:
```
This should work not matter if the default values returned by `get_incident_name()` is `None` or `""`.

File: backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py [hunk 17]
```diff
@@ -1,5 +1,7 @@
 from django.core.management.base import BaseCommand
 
+from common.cachet import create_incident, CachetException, delete_incident, \
```
Comment (Feverfew) [inline backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py:3]: Same here about PEP8

File: backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py [hunk 18]
```diff
@@ -9,8 +11,23 @@ class Command(BaseCommand):
...
-        cache.feed_cache(full=True)
-        print("Done!")
+        try:
+            print("Running OccupEye Caching Operation")
+            print("[+] Feeding Cache")
+            cache = OccupeyeCache()
+            cache.feed_cache(full=True)
+            print("Done!")
+            incident_name = get_incident_name("Occupeye")
+            if incident_name is not None:
```
Comment (Feverfew) [inline backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py:21]:
```suggestion
            if incident_name:

```

Comment (Feverfew) [reply backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py:21]: and maybe an `else`? If this slips by we wouldn't know... Maybe we should print something in this case...

File: backend/uclapi/workspaces/management/commands/feed_occupeye_cache_mini.py [hunk 19]
```diff
@@ -9,8 +11,23 @@ class Command(BaseCommand):
...
-        print("Done!")
+        try:
+            print("Running Mini OccupEye Caching Operation")
+            print("[+] Feeding Cache")
+            cache = OccupeyeCache()
+            cache.feed_cache(full=False)
+            print("Done!")
+            incident_name = get_incident_name("Occupeye-Mini")
+            if incident_name is not None:
+                delete_incident(incident_name)
```
Comment (Feverfew) [inline backend/uclapi/workspaces/management/commands/feed_occupeye_cache_mini.py:22]: ditto here for what I said in `update_gencache.py`

File: backend/uclapi/workspaces/occupeye/cache.py [hunk 20]
```diff
@@ -122,7 +122,7 @@ def cache_survey_data(self):
                 location_data = self._const.SURVEY_LOCATIONS[survey["Name"]]
             else:
                 location_data = {
-                 "lat": "", "long": "", "address": ["", "", "", ""]
+                    "lat": "", "long": "", "address": ["", "", "", ""]
```
Comment (Feverfew) [inline backend/uclapi/workspaces/occupeye/cache.py:125]: unrelated change?

File: backend/uclapi/timetable/management/commands/update_gencache.py (hunk 14 above)
Comment (zipy124) [reply backend/uclapi/timetable/management/commands/update_gencache.py:27]: I don't believe we should follow PEP8 here, since most tool-chains will complain, if you try do the brackets instead of the backslash in an IDEA IDE for example like pycharm you will get an error.

:shrug: ?

File: backend/uclapi/timetable/management/commands/update_gencache.py (hunk 16 above)
Comment (zipy124) [reply backend/uclapi/timetable/management/commands/update_gencache.py:382]: I opted to go for the more verbose method since then the programmer knows exactly what we are expecting back in the default case.

File: backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py (hunk 18 above)
Comment (zipy124) [reply backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py:21]: Can do, it will be kind of annoying since it will always print the else in a dev environment though. Added.

File: backend/uclapi/workspaces/occupeye/cache.py (hunk 20 above)
Comment (zipy124) [reply backend/uclapi/workspaces/occupeye/cache.py:125]: This would be the hound auto-commit function formatting the entire file, this is visual indentation PEP8 compliance.

File: backend/uclapi/timetable/management/commands/update_gencache.py [hunk 21]
```diff
@@ -345,6 +354,24 @@ def handle(self, *args, **options):
...
+            else:
+                print("Could not find appropriate incident in Cachet!")
+
+            if settings.UCLAPI_DOMAIN_CURRENT == "staging.ninja":
+                delete_incident("Gencache-Staging")
+            elif settings.UCLAPI_DOMAIN_CURRENT == "uclapi.com":
+                delete_incident("Gencache-Prod")
+
             call_command('trigger_webhooks')
         except Exception as gencache_error:
```
Comment (Feverfew) [inline backend/uclapi/timetable/management/commands/update_gencache.py:376]: The same comment in `feed_occupeye_cache.py` regarding use of `try/except/else` applies here as well. Unfortunately, unlike in the `try` for occupeye, there's going to have be a lot of code in the `try` block AFAICT, but at least we can move the incident handling to the `else`.

File: backend/uclapi/common/cachet.py [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+and general documentation at
+https://docs.cachethq.io/docs
+"""
+
+
+class CachetException(Exception):
+    pass
+
+
+def get_incident_name(base):
```
Comment (Feverfew) [inline backend/uclapi/common/cachet.py:34]: Should we add type hints? Looks like Django will never support them, but I guess it'd be nice to add them to new code that isn't Django specific. Would sphinx pick them up?If any case we use Sphinx style (it allows many different styles, right?) doc strings (for argument descriptions).

File: backend/uclapi/timetable/management/commands/update_gencache.py [hunk 23]
```diff
@@ -1,27 +1,37 @@
...
-    Modulegroups, ModulegroupsA, ModulegroupsB, \
-    Sites, SitesA, SitesB, \
-    Stuclasses, StuclassesA, StuclassesB, \
-    Stumodules, StumodulesA, StumodulesB, \
-    Students, StudentsA, StudentsB, \
-    Timetable, TimetableA, TimetableB, \
-    Weekmapnumeric, WeekmapnumericA, WeekmapnumericB, \
-    Weekmapstring, WeekmapstringA, WeekmapstringB, \
-    Weekstructure, WeekstructureA, WeekstructureB, \
+
```
[the remark this answers is not in the document]
Comment (zipy124) [reply backend/uclapi/timetable/management/commands/update_gencache.py:2]: That was me doing that manually, to comply with pep8 brackets instead of \?

File: backend/uclapi/common/cachet.py [hunk 24]
```diff
@@ -0,0 +1,151 @@
...
+    """
+    Get and append the appropriate suffix for the component using the current
+    domain. For example if we are running on staging.ninja we add the
+    '-staging' suffix.
+
+    :param base: Base incident name
+    :type base: str
```
Comment (Feverfew) [inline backend/uclapi/common/cachet.py:41]: Yeah I think in this case type hints would be more future proof for `:type var:` and `:rtype:`

File: backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py [hunk 25]
```diff
@@ -9,8 +13,27 @@ class Command(BaseCommand):
...
+            print("[+] Feeding Cache")
+            cache = OccupeyeCache()
+            cache.feed_cache(full=True)
+            print("Done!")
+            incident_name = get_incident_name("Occupeye")
+            if incident_name:
+                delete_incident(incident_name)
+            else:
+                print("No incident present in Cachet!")
+        except Exception as occupeye_error:
```
[the remark this answers is not in the document]
Comment (Feverfew) [reply backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py:34]: @zipy124 correct me if I'm wrong but I don't think this has been addressed at all in any of the functions mentioned?

[the remark this answers is not in the document]
Comment (zipy124) [reply backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py:34]: So it would appear I am an idiot and completely mis-read this review comment and my code.... Sorry after doing the Systems exam paper my brain was frazzeld! Will adress.

**Comment by zipy124** [comment]
type hints added.

In relation to "correct me if I'm wrong but I don't think this has been addressed at all in any of the functions mentioned?" I believe it has been addressed?

File: backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (Feverfew) [reply backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py:34]: @zipy124 this is the bit that needs addressing.

**Comment by Feverfew** [comment]
> type hints added.
> 
> In relation to "correct me if I'm wrong but I don't think this has been addressed at all in any of the functions mentioned?" I believe it has been addressed?

I've unresolved the bits that's need addressing, also tagged you in the relevant comment that explains the issue.

File: backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (Feverfew) [reply backend/uclapi/workspaces/management/commands/feed_occupeye_cache.py:34]:
```
print("Running OccupEye Caching Operation")
print("[+] Feeding Cache")
try:
    cache = OccupeyeCache()
    cache.feed_cache(full=True)
except:
    incident_name = get_incident_name("Occupeye")
    if not incident_name:
        print("Could not find appropriate incident in Cachet!")
        return

    try:
        create_incident(occupeye_error, incident_name)
    except CachetException as cachet_error:
        print(f"Failed to create cachet incident. "
                f"Reason: {repr(cachet_error)}")
    except Exception as cachet_error:
        print(f"Unexpected: Failed to create cachet incident. "
                f"Reason: {repr(cachet_error)}")
else:
    incident_name = get_incident_name("Occupeye")
    if not incident_name:
        print("Could not find appropriate incident in Cachet!" )

    try:
        delete_incident(incident_name)
    except CachetException as cachet_error:
        print(f"Failed to delete cachet incident. "
                f"Reason: {repr(cachet_error)}")
    except Exception as cachet_error:
        print(f"Unexpected: Failed to delete cachet incident. "
                f"Reason: {repr(cachet_error)}")

print("Done!")
```
